Add emergent consensus to sigop and transaction size

gandrewstone opened the pull request:
For non-excessive blocks, Sigops are "limited" to < 20000 (core's number) for blocks < 1MB.  For blocks > 1MB the sigops is a configurable number (net.excessiveSigOp, default 20000) expressed in an amount per rounded-up block size in megabytes.  For example, a 1,000,001 byte block can contain 2*net.excessiveSigOp sigops before it is designated as "excessive".

Its actually more succinctly explained in code:
uint64_t blockMbSize = 1+((blockSize-1)/1000000); 
if (nSigOps > sigOpsPerMb.value*blockMbSize) { EXCESSIVE(); }

sighashes are limited by limiting transaction size.  For compatibility with today's blockchain, any transaction size is allowed in block <= 1MB.  In blocks whose size is > 1MB, a block is excessive if it contains a transaction with a length >= net.excessiveTxn (which is 100KB by default).

gavinandresen committed:
Allow precise tracking of validation sigops / bytes hashed

Adds a ValidationResourceTracker class that is passed to
CheckInputs() / CScriptCheck() to keep track of the exact number
of signature operations required to validate something, and the
exact number of bytes hashed to compute signature hashes.

Also extends CHashWriter to keep track of number of bytes hashed.

Intended to be the starting point for consensus rule changes
to replace the incorrect, messy sigop counting and size-limiting
rules currently in place.

gandrewstone committed:
Merge branch 'accurate_sighash_tracking' of https://github.com/dgenr8/bitcoin into sigop_sighash

gandrewstone committed:
incremental memory use as block sizes increase by reworking CFileBuffer to resize the file buffer dynamically during reindex.  remove 32MB message limit (shadowing already existing 16x EB limit).  Increase max JSON message buffer size so large RPC calls can be made (like getblocktemplate).  Add configurable proportional soft limit (uses accept depth) to sigops and max tx size

gandrewstone committed:
merge 0.12.1bu changes

gandrewstone committed:
sigop and transaction size limits and tests

deadalnix commented on src/hash.h:
```diff
@@ -132,15 +132,17 @@ class CHashWriter
 {
 private:
     CHash256 ctx;
+    size_t nBytesHashed;
```
I would argue against adding this. We should move toward getting rid of this kind of rule rather than working toward helping adding more.

Also, the implementation is poor because ctx already track the amount of byte hashed - it is required as part of the finalization step of sha256.

dgenr8 replied:
The byte counts are not used in the rules.  @gandrewstone just included them for reporting.

deadalnix replied:
Still ctx already count them - check the source if you aren't convinced - so this is pure duplication. And it should go in its own PR.

gandrewstone replied:
yes, thanks for this observation.  Semantically there is exactly one difference between the 2 variables which is that nBytesHashed is not set to 0 during a "Reset" call.  Luckily nobody is using that "feature" so we can "refine" the definition of GetNumBytesHashed to be "num bytes hashed between resets".  Anyway, fixed!

deadalnix commented on src/main.cpp:
```diff
@@ -414,6 +420,14 @@ bool MarkBlockAsReceived(const uint256& hash) {
         LogPrint("thin", "Received block %s in %.2f seconds\n", hash.ToString(), nResponseTime);
         LogPrint("thin", "Average block response time is %.2f seconds\n", avgResponseTime);
+        if (maxBlocksInTransitPerPeer.value != 0)
+	  {
```
Identitation is screwed up.

deadalnix commented on src/txmempool.h:
```diff
@@ -86,6 +86,8 @@ class CTxMemPoolEntry
     bool spendsCoinbase; //! keep track of transactions that spend a coinbase
     unsigned int sigOpCount; //! Legacy sig ops plus P2SH sig op count
+    uint64_t runtimeSigOpCount; //! Runtime signature operation count
+    uint64_t runtimeSighashBytes; //! Runtime bytes hashed for signature operations
```
See comment in the sighash. I don't think we should be adding this.

deadalnix commented on src/unlimited.cpp:
```diff
@@ -693,12 +693,42 @@ int isChainExcessive(const CBlockIndex* blk, unsigned int goBack)
+      {
+	// Check transaction size to limit sighash
+        if (largestTx > maxTxSize.value)
+          {
```
I know the indentation hasn't been very consistent but still. This has 5 indentation levels in 5 lines, plus the indentation is inconsistent with other construct in the same file. For instance you indent { by 2 spaces after if.

ftrader replied:
Has anyone tried the contrib/devtools/clang-format.py script?
We have all those nice settings in src/.clang-format.

gandrewstone replied:
thanks!  Its the tabs vs spaces issue again, I fixed to spaces.  I will clang-format the BU-specific files before release, in a separate check in.

deadalnix commented on src/unlimited.cpp:
```diff
@@ -693,12 +693,42 @@ int isChainExcessive(const CBlockIndex* blk, unsigned int goBack)
+
+        // Check max sigops
+        if (nSigOps > BLOCKSTREAM_CORE_MAX_BLOCK_SIGOPS)
+	  {
```
But here is is one space.

gandrewstone replied:
tabs vs spaces again, fixed

deadalnix commented on src/main.cpp:
```diff
@@ -4345,15 +4409,15 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, CDiskB
                 uint256 hash = block.GetHash();
                 if (hash != chainparams.GetConsensus().hashGenesisBlock && mapBlockIndex.find(block.hashPrevBlock) == mapBlockIndex.end()) {
-                    LogPrint("reindex", "%s: Out of order block %s, parent %s not known\n", __func__, hash.ToString(),
+		  LogPrint("reindex", "%s: Out of order block %s (created %s), parent %s not known\n", __func__, hash.ToString(), DateTimeStrFormat("%Y-%m-%d",block.nTime),
```
2 tabs and 2 spaces...

gandrewstone replied:
arrrg!

deadalnix commented on src/txmempool.h:
```diff
@@ -117,6 +119,8 @@ class CTxMemPoolEntry
     bool WasClearAtEntry() const { return hadNoDependencies; }
     unsigned int GetSigOpCount() const { return sigOpCount; }
+    uint64_t GetRuntimeSigOpCount() const { return runtimeSigOpCount; }
+    uint64_t GetRuntimeSighashBytes() const { return runtimeSighashBytes; }
```
dito

deadalnix commented:
So I agree with the general algorithm. however, this PR seems like it contains a lot of changes that are unrelated.

For instance, it introduce sighash byte counting, a remake of CBufferedFile, various changes in the way stats are working. This is very hard to review in its current form. I would advocate for this to be spited in several PR that are focused on doing one thing. That would be much easier to review, much easier to revert if something goes wrong, and much easier for anyone digging in the history to understand what happens.

dagurval commented on src/globals.cpp:
```diff
@@ -135,6 +137,21 @@ CTweakRef<unsigned int> briTweak("net.blockRetryInterval","How long to wait in m
+/** Size of the "block download window": how far ahead of our current height do we fetch?
+ *  Larger windows tolerate larger download speed differences between peer, but increase the potential
+ *  degree of disordering of blocks on disk (which make reindexing and in the future perhaps pruning
+ *  harder). We'll probably want to make this a per-peer adaptive value at some point. */
```
Same comment for both _net.blockDownloadWindow_ and _reindex.typicalBlockSize_

gandrewstone replied:
thanks

dagurval commented on src/main.cpp:
```diff
@@ -3561,10 +3572,15 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
 
     uint64_t nSigOps = 0;
     uint64_t nTx = 0;  // BU: count the number of transactions in case the CheckExcessive function wants to use this as criteria
+    uint64_t nTxSize = 0;  // BU: track the longest transaction
```
Nit: nTxSize is only used in the scope of a for loop, so declare it there?

gandrewstone replied:
done

ftrader commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -115,6 +115,8 @@
 
 testScriptsExt = [ RpcTest(t) for t in [
+    'txPerf',
+    'excessive --extended',
```
Generally, I would favor separate tests for testing separate functionality unless the "extended" is used to denote more iterations or range of an input.
This PR adds separate functionality to the "excessive" category, namely a sigops criterion. It makes sense to me to test this separately, as this simplifies into smaller individual test scripts with specific focus.
It also makes it easier to see what's going on when looking at the list of tests. It took me quite a while to parse the new structure of excessive.py to see how self.extended was passed in and for what it was used.
With separate tests this would be much easier to discern immediately.

gandrewstone replied:
Its mostly more iterations.  In one case I skipped a test in the non-extended version since the test takes a long time and can't be shortened.  It has to create 1MB+ of transactions and then create and sign a large tx.  This takes time.

dagurval commented on src/streams.h:
```diff
@@ -591,6 +770,47 @@ class CBufferedFile
     }
+
+    // if the current buffer doesn't have amt more data, then extend it by that much
+    bool GrowTo(uint64_t amt)
```
Nit: This method always returns true. Might as well be void.

gandrewstone replied:
thx fixed

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
 
 class ExcessiveBlockTest (BitcoinTestFramework):
+    def __init__(self,extended=False):
+      self.extended = extended
```
As mentioned above, consider splitting costly testing into a separate script.

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
-    def run_test (self):
-        BitcoinTestFramework.run_test (self)
+    def setup_network(self, split=False):
+        self.nodes = start_nodes(4, self.options.tmpdir,timewait=60*10)
```
Where does the magic 60*10 value come from?

gandrewstone replied:
The time to wait for a RPC response doesn't really matter, except that it needs to be longer than the default 30 seconds to sign a 100k+ transaction.  In this case 10 minutes is how long travis is willing to wait... so not much point in waiting longer than that.

ftrader replied:
it's possible to override Travis wait time ... (just saying, it doesn't seem much use if total time is limited to 50min)

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+        self.is_network_split=False
+        self.sync_all()
+
+        if 0:  # Use to create wallet with lots of addresses
```
delete unused code

gandrewstone replied:
I added a much more descriptive comment that explains exactly why I want to leave this as if 0

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+    def run_test(self):
+      BitcoinTestFramework.run_test (self)
+
+      # exact point in chain is not required
```
don't leave commented-out code unless with a comment explaining why its been deactivated (temporarily)

gandrewstone replied:
ok

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+      #assert_equal (tips[0]['status'], 'active')
+
+      self.testExcessiveBlock()
+      self.testExcessiveTx()
```
testExcessiveTx is imo a candidate for a separate test script which focuses on the tx aspect

gandrewstone replied:
hmm... to make 2 files or 2 routines.  This is a qualitative judgement...

ftrader replied:
File length & class/method length are quantitative ones which favors splitting up into smaller modules.
One could argue the excessive.py is getting excessively long :-)

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+    def testExcessiveTx(self):
+      TEST_SIZE=20
+      logging.info("Test excessive transactions")
+      if 1:
```
seeing `if 1:` and `if 0:` in code to be merged bothers me greatly
not only is it superfluous, it hinders readability

gandrewstone replied:
hmm... I think that it enhances readability by blocking out code sections.

ftrader replied:
Above you've stated the only reason you do this is to ease merging from upstream. This doesn't apply here.

The '0's are dead code, pure and simple, wasting my time as a reviewer, introducing more possibility for mistakes during review and further development etc.

Nearly every coding guide says get rid of dead code unless it AAABSOLUTELY needs to stick around. That can't be the case for this test which is unique to BU.

Like another person said: there's version control for that...

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
-        assert_equal (tips[0]['branchlen'], 0)
-        assert_equal (tips[0]['height'], 200)
-        assert_equal (tips[0]['status'], 'active')
+        #assert_equal (len (tips), 1)
```
remove?

gandrewstone replied:
yes thx

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+        if 1:
+  	  logging.info("Creating addresses...")
+	  addrs = [ self.nodes[0].getnewaddress() for _ in range(TEST_SIZE+1)]
+        else:
```
It seems you want a parameter to run this. An `if 1: ... else:` doesn't make sense any other way.

gandrewstone replied:
I added comment.  It allows pre-generation of wallet addresses, but requires some manual steps so a parameter would just be confusing.  The code in "if 0:" is basically notes left to the next engineer describing shortcuts or debugging techniques that may be useful.

ftrader replied:
It's dead code. As the next engineer (in this case, reviewer), I don't want that around.
Please put your notes in commit messages, inline comments or developer documentation (README, or whatever).

And for the love of *, don't make me modify 0's to 1's in N places in the code to get some debugging feature activated. Create a debugging parameter in the module or something (what's wrong with a parameter - it's nowhere near as confusing as these uncommented `if 0:` / `if 1:` sections.

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+      self.testExcessiveBlock()
+      self.testExcessiveTx()
+
+    def testExcessiveTx(self):
```
This ought to have a docstring describing, in plain English, what this test is supposed to do.
Same goes for any other non-trivial methods in the class, and the class itself, if other parts are lacking these too.

gandrewstone replied:
ok

deadalnix commented on src/streams.h:
```diff
@@ -463,9 +473,15 @@ class CBufferedFile
+    //std::vector<char> vchBuf; // the buffer
+    //enum { CHUNK_SIZE = 200000 };  // BU how much additional to allocate if forced to resize
 protected:
+#if 0
```
I see that you often comment out code or ifdef it out. Is there a reason ? All history is kept in git, so, as long as PR are kept on point, it is easy to find what was the previous version of the code.

gandrewstone replied:
Yes, we need to "additively remove" code.  Otherwise during a rebase, it looks to me like Core has recently added it.  By adding the #if 0, the merge works without my intervention, even if Core modifies the Seek() function.

I don't like it much, but we are stuck with it for now.

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+            count+=1
+            utxo = wallet.pop()
+            outp = {}
+	    outp[addrs[count%len(addrs)]] = utxo["amount"]
```
I wonder if this is just Github rendering the indentation funny, or how is it that the Python interpreter doesn't fall over this seemingly inconsistent indenting?

gandrewstone replied:
yes, they are tabs that are somehow in the code.  I have added tab removal to my editor now, hopefully that will solve the problem in the future.  Thanks for spotting these

ftrader replied:
These weren't the only places - I just didn't comment on all of them because too many. Please look over the file, and run it through tools like PEP8 or look at it in Eclipse PyDev (which immediately shows trouble spots in nice color blips along the editor margin).

sickpig replied:
@gandrewstone I know you are using emacs so this is could worth considering in regards of indentation/formatting: https://www.emacswiki.org/emacs/IndentingPython 
The same apply for C++ code: https://www.emacswiki.org/emacs/CPlusPlusMode#toc1

gandrewstone replied:
I actually AM using python and c modes... I think I've chased it down to a config option:
(setq-default indent-tabs-mode nil)

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+            if n >= len(addrs): n=0
+	
+          self.nodes[0].generate(1)
+	  self.sync_all()
```
Weird indenting.
https://www.python.org/dev/peps/pep-0008/#indentation

gandrewstone replied:
Thanks!  looks like a tab vs spaces issue actually.  I'll fix

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+          latest = counts[0]
+          assert_equal(counts, [latest,latest,latest,latest]) # Verify that all nodes accepted the block, even if some of them didn't have the transaction.  They should all accept a <= 1MB block with a tx <= 1MB
+
+  	  # mafter = [ (lambda y: (y["size"],y["bytes"]))(x.getmempoolinfo()) for x in self.nodes]
```
dead code (debugging?)

gandrewstone replied:
I'll delete... its more just some interesting info that could be printed out by the next developer...

dagurval commented on src/streams.h:
```diff
@@ -442,12 +444,20 @@ class CAutoFile
     }
 };
 
+#if 0
```
Why add all this dead code (dead because its inside #if 0)) ?

gandrewstone replied:
thanks my bad -- the was a different approach that I chose not to use...

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+          mpool = [ (lambda y: (y["size"],y["bytes"]))(x.getmempoolinfo()) for x in self.nodes]
+          print mpool
+          largeBlock = self.nodes[0].generate(1)
+          time.sleep(10) # can't sync b/c nodes won't be in sync
```
Fixed sleeps are bad news for tests under most circumstances.
I see this one repeats a couple of times further down as well.
Is there a way this can be made into a while loop that checks for a successful sync condition, like you did elsewhere in the script?
This would be much more robust and give the tester a much better idea of what went wrong if the test gets stuck on syncing rather than failing due to an elapsed timeout (which could have a myriad reasons).

gandrewstone replied:
ok

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+          # assert(binascii.unhexlify(tx) > 100000) # txn has to be big for the test to work
+	  origCounts = [ x.getblockcount() for x in self.nodes ]
+          mpool = [ (lambda y: (y["size"],y["bytes"]))(x.getmempoolinfo()) for x in self.nodes]
+          print mpool
```
print vs logging.\* ?

gandrewstone replied:
thx

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+            counts = [ x.getblockcount() for x in self.nodes ]
+            if counts[0] == counts[1]:  # this is a large block with lots of tx so can take a LONG time to sync on one computer
+              break
+            time.sleep(10) # can't sync b/c nodes won't be in sync
```
see my comment on fixed sleeps above
the while: logic used here is repeated twice with almost the same loop contents (only node numbers used for comparisons differ).
Consider factoring this sync loop out into a method.

gandrewstone replied:
good idea

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -40,116 +254,160 @@ def run_test (self):
+        base = counts[0]
+        logging.info("node0")
         self.nodes[0].generate(1)
         time.sleep(2) #give blocks a chance to fully propagate
```
see my comment on fixed sleeps above. As this is old code (and the pattern repeats in the historic code in this test, consider perhaps a sync method which can help)

gandrewstone replied:
ok in this case I want to give them the chance to propagate but I expect that the propagation will fail.  So there's nothing to wait for that is detectable anyway

ftrader commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -99,11 +108,31 @@ def sync_mempools(rpc_connections, wait=1):
+      shutil.copyfile(wallet,os.path.join(regtestdir, "wallet.dat"))
+      
+
+    if 0:
```
Please no debugging sections like this in framework modules.
It makes reviewing tougher as well.
If you need it, put it under some controllable debug parameters.

gandrewstone replied:
yes, these were the original params but should be removed

dagurval commented on src/main.cpp:
```diff
@@ -3505,17 +3516,17 @@ bool CheckBlockHeader(const CBlockHeader& block, CValidationState& state, bool f
 }
 
-bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bool fCheckMerkleRoot, bool fConservative)
+uint64_t CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bool fCheckMerkleRoot, bool fConservative)
```
Why does CheckBlock return block size some times?

gandrewstone replied:
thanks!  I thought I needed to return the block size or recalculate it, but it turns out its in nBlockSize.  I'll revert these

ftrader commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -228,7 +260,7 @@ def start_node(i, dirname, extra_args=None, rpchost=None, timewait=None, binary=
         binary = os.getenv("BITCOIND", "bitcoind")
     # RPC tests still depend on free transactions
-    args = [ binary, "-datadir="+datadir, "-server", "-keypool=1", "-discover=0", "-rest", "-blockprioritysize=50000" ]
+    args = [ binary, "-datadir="+datadir, "-keypool=1", "-rest"] # , "-blockprioritysize=50000" ]
```
Remove dead code. Maybe leave a comment if you think something about blockprioritysize can influence the test, or not.

gandrewstone replied:
This was hard-coded in there.  I don't think that its "right" to run all the tests with nonstandard flags.  I'm left it commented so we can have a conversation about why this was put in in the first place...

ftrader replied:
ok, good point.

ftrader commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -292,6 +324,15 @@ def connect_nodes_bi(nodes, a, b):
+    """Connect every node in this list to every other node in the list"""
+    for frm in nodes:
+      for to in nodes:
+        if frm==to: continue
```
surround binary operators with spaces
https://www.python.org/dev/peps/pep-0008/#id28

gandrewstone replied:
? not seeing the recommendation

ftrader replied:
`if frm==to` <-- hard to read

`if frm == to`   <--- much easier

gandrewstone replied:
hmm... I guess my problem is that I started with python before PEP008.  Anyway, I'll defer to your more exacting standards :-)

ftrader commented on qa/rpc-tests/txPerf.py:
```diff
@@ -0,0 +1,182 @@
+              pdb.set_trace()
+              continue
+
+            time.sleep(4) # give the transaction time to propagate so we generate tx validation data separately from block validation data
```
see earlier comment about fixed sleeps

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+
+          logging.info("Syncing node2")
+          largeBlock3 = self.nodes[0].generate(1)
+          while 1:
```
refer to my earlier comment about factoring out such sync loops

gandrewstone replied:
done

ftrader commented on qa/rpc-tests/excessive.py:
```diff
@@ -12,17 +12,231 @@
+
+          logging.info("Syncing node3")          
+          largeBlock4 = self.nodes[0].generate(1)
+          while 1:
```
refer to my earlier comment about factoring out this sync loop

gandrewstone replied:
check

ftrader commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -41,9 +41,9 @@ def run_test(self):
         pass
 
-    def setup_chain(self):
+    def setup_chain(self,bitcoinConfDict=None, wallets=None):
```
these new parameters need some brief introduction, here or in another suitable place (docstring)

gandrewstone replied:
done

dgenr8 commented:
This should be a defensible strategy.  I don't see a problem with a default excessive transaction size of 100K. The overhead to create another such transaction is 31 bytes or .031% (another 13-byte header plus two 9-byte counts).

ftrader commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -95,7 +104,7 @@ def join_network(self):
         self.setup_network(False)
 
-    def main(self):
+    def main(self,argsOverride=None,bitcoinConfDict=None,wallets=None):
```
these new parameters need some brief introduction, here or in another suitable place (docstring)

gandrewstone replied:
roger roger

ftrader commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -99,11 +108,31 @@ def sync_mempools(rpc_connections, wait=1):
     if not os.path.isdir(datadir):
         os.makedirs(datadir)
+    
+    defaults = {"server":1, "discover":0, "regtest":1,"rpcuser":"rt","rpcpassword":"rt",
```
With much structural changes going on, I'm not even sure if "blockprioritysize":2000000 is a newly added default or not.
It seemed to be, based on the addition in the commented out (`if 0:`) section elsewhere.

If it is truly an addition, please comment why this parameter is needed and why this value was chosen.

gandrewstone replied:
thanks, and good catch.  The original code hard-coded the bitcoin.conf file contents within the framework.  I realised that tests might need custom config (I do), so I added the whole config parameters.  But I missed moving this parameter out to the new test-specific config.

ftrader replied:
I find it really useful when there's some justification (by comments in code) as to why certain values are used - whether in a test or elsewhere...

gandrewstone replied:
ok added a comment (and moved the parameter to excessive.py)

ftrader commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -398,6 +439,83 @@ def random_transaction(nodes, amount, min_fee, fee_increment, fee_variants):
+	  else:
+	      outp[str(a)] = float(amtPer/BTC)
+
+    #      try:
```
moved up, I suppose?

gandrewstone replied:
I don't understand the exact question, but I pulled this code from prior code I wrote and I clearly see it needs a lot of cleanup so I cleaned it up.

ftrader replied:
was referring to the commented-out 'try' hanging around. I presumed it had been moved upwards, but either way, seems like the commented-out code could just be removed.

gandrewstone replied:
ok good I cleaned that try up.

ftrader commented on qa/rpc-tests/txPerf.py:
```diff
@@ -0,0 +1,182 @@
+import time
+import math
+import json
+from decimal import *
```
wildcard import really needed?

gandrewstone replied:
is it considered bad form?

ftrader replied:
> Wildcard imports ( from <module> import \* ) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools. There is one defensible use case for a wildcard import, which is to republish an internal interface as part of a public API (for example, overwriting a pure Python implementation of an interface with the definitions from an optional accelerator module and exactly which definitions will be overwritten isn't known in advance). 

https://www.python.org/dev/peps/pep-0008/

NOTE: there is a 'pep8' command line utility which does a nice job of showing up violations of PEP8. I highly recommend it.

ftrader commented on src/consensus/consensus.h:
```diff
@@ -8,10 +8,15 @@
 
 /** The maximum allowed size for a serialized block, in bytes (network rule) */
-static const unsigned int BU_MAX_BLOCK_SIZE = 32000000;  // BU: this constant is deprecated but is still used in a few areas such as allocation of memory.  Removing it is a tradeoff between being perfect and changing more code. TODO: remove this entirely
+//static const unsigned int BU_MAX_BLOCK_SIZE = 32000000;  // BU: this constant is deprecated but is still used in a few areas such as allocation of memory.  Removing it is a tradeoff between being perfect and changing more code. TODO: remove this entirely
```
Seems this constant is being removed entirely (great!)
Why is the commented out code left behind?

gandrewstone replied:
Its the MAX_BLOCK_SIZE in Core, so when I rebase by "additively removing" it, the rebase works cleanly.

ftrader reviewed (commented):
A more in-depth review is needed to determine if the extensions to the tests cover the necessary cases.

dgenr8 committed:
Allow precise tracking of validation sigops/bytes

Adds a ValidationResourceTracker class that is passed to
CheckInputs() / CScriptCheck() to keep track of the exact number
of signature operations required to validate something, and the
exact number of bytes hashed to compute signature hashes.

Also extends CHashWriter to keep track of number of bytes hashed.

This is a subset of the original work by Gavin Andresen.

gandrewstone committed:
cleanup formatting, etc

gandrewstone committed:
remove duplicate sighash byte count

gmaxwell commented:
ACK.  This change will make it easier for me to explain the profound insanity of this project.

deadalnix commented:
The sighash thing looks good. I think a test for it would ensure that this returns the expected result and isn't broken unexpectedly in the future. Why don't you open a PR for that and rebase this one on top of it ?

Keeping things focused makes for history that is easy to understand. Merging branches with a bunch of commit that do and undo various stuff, on the other hand, makes for very confusing history - And I can attest this first hand as I have been digging in history quite a lot to figure out what is going on lately, but it is a huge mess. This also makes various things harder to do, such as bisecting.

dgenr8 commented:
@gandrewstone No problem with the byte count change. d9b8928 should also be replaced by the updated c902cc4 from #122.  This will reduce the changed file count by 2 and eliminate any possibility of the runtime counts affecting block validity.

dgenr8 commented:
Shouldn't ``net.excessiveSigOp`` and ``net.excessiveTxn`` be part of the coinbase string? Otherwise, how will anyone know whether we will reject their block?

dgenr8 commented:
``net.excessiveTxn`` and ``net.excessiveSigOp`` can cause us not to accept other miners' blocks.  Don't they need to be communicated in the coinbase message?

gandrewstone committed:
Merge branch '0.12.1bu' of https://github.com/BitcoinUnlimited/BitcoinUnlimited into sigop_sighash

gandrewstone committed:
merge

gandrewstone committed:
merge

gandrewstone committed:
extended testing and sigops proportional limits

gandrewstone committed:
merge

gandrewstone committed:
add some prints so travis knows that the test is still going

gandrewstone committed:
merge

gandrewstone committed:
move long running tests to extended

gandrewstone committed:
in block generation, account for block header and largest possible coinbase size.  Clean up excessive tests

ftrader commented:
Is this PR the new consensus code for BUIP040?

zander commented:
@ftrader when I asked on slack a week ago, thezerg confirmed your suspicion.

sickpig commented:
@ftrader, @tomz 

Just for the sake of completeness This is complete response from @gandrewstone (thezerg)

_[...] as per the Articles, the BUIP controls the official release process, not what is checked in.   Also the BUIP is not linked to a particular PR, it covers specific features.  This commit has a lot of useful stuff, like enhancements to the test framework and changes to support very large blocks.  If the BUIP does not pass, its just a few LOC to remove the features covered in the BUIP.
Finally, from a operational standpoint, I want to get any largish PR merged so I can run clang-format over the BU-specific files today without having a subsequent merge headache._

zander commented:
Thanks @sickpig, I didn't feel it was my place to paste a message shared on a private board, great that you are able to shared that.

ftrader commented:
Thanks sickpig, that clarifies it for me.
